core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -257,17 +257,17 @@ private static class Task<T> implements Supplier<Optional<Task<T>>>, Closeable {
     @Override
     public Optional<Task<T>> get() {
       try {
+        if (queue.size() >= approximateMaxQueueSize) {
```

Comment:
https://github.com/apache/iceberg/pull/11781/files#diff-0586c055bbd9d220727cfa60c802321693a959c0cfb3a510376fc5a5bb1dc897R99

`this.taskFutures = new CompletableFuture[2 * ThreadPools.WORKER_THREAD_POOL_SIZE];`

here the global thread pool size is driving the taskFuture array size, can this leads to more tasks then the available S3 connections?

as I believe each table scan will create a new `ParallelIterable` ?

Reply:
> here the global thread pool size is driving the taskFuture array size, can this leads to more tasks then the available S3 connections?

Correct

> as I believe each table scan will create a new ParallelIterable ?

Yes.

Reply:
so, each table scan leads to `2*cores` tasks/s3 connections?
and still can exhaust S3 connections?

Reply:
> so, each table scan leads to 2*cores tasks/s3 connections?

potentially, yes. but there are extra conditions like queue must be full.

Reply:
Are we saying if queue is full then only we can reach to a condition that can lead to s3 connections are all exhausted?
Though adding task to queue does not mean a S3 connection?

Reply:
In order for connection to be kept around task has to be started and then it needs to yield. If task doesn't yield then the S3 connection is fully utilized and returned afterwards